Parsimony/Parser.cs

```diff
@@ -0,0 +1,212 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Parsimony
+{
+    /// <summary>
+    /// A CLI parser for <typeparamref name="T"/>
+    /// </summary>
+    /// <typeparam name="T">The type of the options to parse.</typeparam>
+    public class Parser<T> where T : notnull, new()
```

Comment:
Generally I loathe shared state but in this case given that you've got a bunch of little object that understand their own types wouldn't it be a bit better from a type safety/responsibility perspective to delegate the per token state manipulation to those objects or children of those objects than it is to handle it in a parent that has to reflect all of that out? 


ex
```
parser -> Option -> StateHandler

TokenStream -> Option1 -> Match ? -> StateHandler
TokenStream[1..] -> Option1 -> Match? No -> Option2 -> StateHandler
   
```

Seems like it would end up with less reflection magic and more type safety overall

Reply:
plus I think it segregates the responsibilities a bit more

Reply:
I considered having each `Option` be its own parser. The only thing I didn't like about it was that if the options are their own parsers then they have the ability to know whether they're suitable for parsing the next token/s in a given stream. AFAICS this means you either
- test each parser at each state until one works (seems wasteful), or
- inspect the steam to chose the parser, then use a parser that could or does re-evaluate its own suitability

As I'm thinking about it now though, the top level one could always just use the name token to select the parser then pass it a parse state with the option name removed and have the option parser parse the value token 🤔 

I'll revisit

Reply:
Yeah I think the merits come from the ability to avoid a lot of type erasure and punning and if it works out like I think would make it feel less like we should just write it in python :P

Reply:
write the option parser in python and embed a python interpretter in the dotnet apps lol

Reply:
TIHI